lib/src/package_config_impl.dart

```diff
@@ -375,122 +395,124 @@ abstract class PackageTree {
+      though.
+      // 3) The new package is inside the packageUriRoot of existing package.
+      if (_beginsWith(0, existingPackage.packageUriRoot.toString(),
+          newPackage.root.toString())) {
+        onError(ConflictException(
+            newPackage, existingPackage, ConflictType.insidePackageRoot));
+        return true;
+      }
+      */
     }
+    return false;
   }
 
-  /// Tries to (add) `package` to the tree.
+  /// Tries to add `newPackage` to the tree.
   ///
   /// Reports a [ConflictException] if the added package conflicts with an
   /// existing package.
-  /// It conflicts if its root or package root is the same as another
-  /// package's root or package root, or is between the two.
+  /// It conflicts if its root or package root is the same as an existing
+  /// package's root or package root, is between the two, or if it's inside the
+  /// package root of an existing package.
   ///
-  /// If a conflict is detected between [package] and a previous package,
+  /// If a conflict is detected between [newPackage] and a previous package,
   /// then [onError] is called with a [ConflictException] object
-  /// and the [package] is not added to the tree.
+  /// and the [newPackage] is not added to the tree.
   ///
   /// The packages are added in order of their root path.
-  /// It is never necessary to insert a node between two existing levels.
-  void add(
-      int start, SimplePackage package, void Function(Object error) onError) {
-    var path = package.root.toString();
-    for (var treePackage in packages) {
-      // Check is package is inside treePackage.
-      var treePackagePath = treePackage.root.toString();
-      assert(treePackagePath.length > start);
-      assert(path.startsWith(treePackagePath.substring(0, start)));
-      if (_beginsWith(start, treePackagePath, path)) {
-        // Package *is* inside treePackage.
-        var treePackagePathLength = treePackagePath.length;
-        if (path.length == treePackagePathLength) {
-          // Has same root. Do not add package.
-          onError(ConflictException.root(package, treePackage));
-          return;
-        }
-        var treePackageUriRoot = treePackage.packageUriRoot.toString();
-        if (_beginsWith(treePackagePathLength, path, treePackageUriRoot)) {
-          // The treePackage's package root is inside package, which is inside
-          // the treePackage. This is not allowed.
-          onError(ConflictException.packageRoot(package, treePackage));
-          return;
-        }
-        _treeOf(treePackage).add(treePackagePathLength, package, onError);
-        return;
+  void add(SimplePackage newPackage, void Function(Object error) onError) {
+    var root = newPackage.root;
+    var currentMapHelper = _map[root.scheme] ??= _TrielikePackageTreeHelper();
+    if (_checkConflict(currentMapHelper, newPackage, onError)) return;
+    var segments = root.pathSegments;
+    for (var i = 0; i < segments.length - 1; i++) {
```

Comment:
There's a small gotcha here. I'm skipping the last segment because the last one is the empty string because root is a directory.
If not skipping it we'd get into trouble, e.g. "a/b" wouldn't be "inside" the package "a" as I'll try to demonstrate with this "drawing":
```
                    a
                   /  \
     (empty string)    b
node with package a     \
                         (empty string)
                         node with package b
```